### PR TITLE
fix: Redirect user to login if session is terminated

### LINKED ISSUE #49187: [Bug]: Getting Unexpected error: Invalid response 401 in Files app
### ⚠️ This issue respects the following points: ⚠️

- [x] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [x] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [x] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [x] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

After upgrade from 28 to 29 I've started to get Unexpected error: Invalid response 401 in Files app when trying to open any folder in Files webUI after leaving the webpage for a while. Clicking on Retry is not helpful. The only way is to manually reload the webpage.
Image

### Steps to reproduce

1. Open webUI files
2. Go somewhere
3. Leave it for a while: 3-6-12 hours
4. Click on any folder in the Files app
5. Observe Unexpected error: Invalid response 401 

### Expected behavior

The webpage should transparently renew without manual intrusion of a user, when a user comes back to it, similarly to Google Drive or any other cloud solution.

### Nextcloud Server version

29

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.2

### Web server

Apache (supported)

### Database engine version

MariaDB

### Is this bug present after an update or on a fresh install?

Upgraded to a MAJOR version (ex. 28 to 29)

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```json

```

### List of activated Apps

```shell

```

### Nextcloud Signing status

```shell

```

### Nextcloud Logs

```json

```

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by susnux**
* Resolves: https://github.com/nextcloud/server/issues/49187#issuecomment-2468285370

## Summary
If a session timed out or was closed in another tab, then currently the user gets random error messages. This intercepts 401 responses (should only happen if logged out, or the users does something wrong).

If we get a 401, we make sure its because of the session, by checking if the user can access the files app.
If that is also the case we forward the user to the login page and set the redirect URL to the last used URL.

## Screen recording

[Bildschirmaufnahme_20241111_174951.webm](https://github.com/user-attachments/assets/1ec0902e-97d0-4f36-957c-085795b489bc)

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by artonge — COMMENTED** [review]
Shouldn't this be added to the `@nextcloud/axios` lib?
But is the webdav lib using `@nc/axios`?

**Comment by susnux** [comment]
> Shouldn't this be added to the @nextcloud/axios lib?

No, similar as the other interceptors, this is not a feature developers can opt-in with using nc-axios, but it is a core feature.
Meaning it should work for all requests, even if you just use vanilla JavaScript.

**Review by Pytal — APPROVED** [review]

File: core/src/utils/xhr-request.js [hunk 1]
```diff
@@ -26,6 +27,45 @@ const isNextcloudUrl = (url) => {
...
+
+	try {
+		// We need to check this as a 401 in the first place could also come from other reasons
+		const { status } = await window.fetch(generateUrl('/apps/files'))
+		if (status === 401) {
+			console.warn('User session was terminated, forwarding to login page.')
+			window.location = generateUrl(
+				'/login?redirect_url={redirect}',
+				{
+					redirect: window.location.pathname + window.location.search,
```
[the remark this answers is not in the document]
Comment (susnux) [reply core/src/utils/xhr-request.js:58]: yes makes sense, but I think we need to URL encode the `#`

**Comment by susnux** [comment]
Tests added.

**Review by skjnldsv — APPROVED** [review]

**Comment by inet-cologne** [comment]
@susnux , @skjnldsv , @artonge , @Pytal 
Hey guys,
may it be that the change made here in #49560 or #49208 produces somehow login problems within Nextcloud files app, Nextcloud Talk app and perhaps causing issues with Calendar subscription synchronising?
Since a couple of days after updating (several instances!!) Nextcloud Server to 30.0.5, problems were occurring that Login fails. These could be fixed by deleting and reinstalling and reconfiguring the apps.
30.0.6 does not fix the situation either but now even reinstallation does not help anymore.
**Login with the above mentioned iOS apps are not possible any more!!**
Please see issue #50619!

Thanks for checking that!!
Chris

**Comment by inet-cologne** [comment]
@susnux , @skjnldsv , @artonge , @Pytal
Hey guys, could you pleeeease check this?
Thanks a lot!!

**Pull request merged**